exposure-notification/src/main/java/fi/thl/covid19/exposurenotification/efgs/FederationGatewaySyncProcessor.java

```diff
@@ -5,48 +5,55 @@
 import org.springframework.scheduling.annotation.Scheduled;
 import org.springframework.stereotype.Component;
 
-import java.time.Duration;
 import java.time.LocalDate;
 import java.time.ZoneOffset;
 import java.util.Optional;
 import java.util.Set;
+import java.util.concurrent.atomic.AtomicReference;
 
 import static net.logstash.logback.argument.StructuredArguments.keyValue;
 
 @Component
 public class FederationGatewaySyncProcessor {
 
     private static final Logger LOG = LoggerFactory.getLogger(FederationGatewaySyncProcessor.class);
+    private final AtomicReference<LocalDate> lastInboundSyncFromEfgs;
 
-    private final FederationGatewayService federationGatewayService;
+    private final FederationGatewaySyncService federationGatewaySyncService;
 
-    public FederationGatewaySyncProcessor(FederationGatewayService federationGatewayService) {
-        this.federationGatewayService = federationGatewayService;
+    public FederationGatewaySyncProcessor(FederationGatewaySyncService federationGatewaySyncService) {
+        this.federationGatewaySyncService = federationGatewaySyncService;
+        this.lastInboundSyncFromEfgs = new AtomicReference<>(LocalDate.now(ZoneOffset.UTC));
     }
 
     @Scheduled(initialDelayString = "${covid19.federation-gateway.upload-interval}",
             fixedRateString = "${covid19.federation-gateway.upload-interval}")
     private void runExportToEfgs() {
         LOG.info("Starting scheduled export to efgs.");
-        Set<Long> operationIds = federationGatewayService.startOutbound(false);
+        Set<Long> operationIds = federationGatewaySyncService.startOutbound(false);
         LOG.info("Scheduled export to efgs finished. {}", keyValue("operationId", operationIds));
     }
 
     @Scheduled(initialDelayString = "${covid19.federation-gateway.download-interval}",
             fixedRateString = "${covid19.federation-gateway.download-interval}")
     private void runImportFromEfgs() {
-        LOG.info("Starting scheduled import from efgs.");
-        federationGatewayService.startInbound(LocalDate.now(ZoneOffset.UTC), Optional.empty());
-        LOG.info("Scheduled import from efgs finished.");
+        LocalDate today = LocalDate.now(ZoneOffset.UTC);
+        LocalDate last = lastInboundSyncFromEfgs.get();
```

Comment:
This could be done atomically with updateAndGet.
Also, now that I read it again, I started wondering about storing the integration state in memory like this. Shouldn't it be in the DB? Or how is this meant to behave when the service is restarted?

Reply:
It doesn't really matter in this case. This is just meant to prevent too much downloading and we cannot prevent that entirely if we don't want to go too far with this. Anyway, actual fail safe is on the db-layer. So, it's ok to initialize it again on startup. And all nodes also. We're heading towards callback and when it is going to be activated this logic will be there just for special cases.

Also, in this case it's totally acceptable to let the last update to be written to the atomicreference, even if there is parallel processing involved, because that's what it's trying to representate, when the last operation ran.

About updateAndGet, not so sure about that, at least in case if you mean that we would pass the entire operation for it. There is some chance that updateAndGet would be called twice and that's something what we're trying to prevent in here. However, it might possible to use it at start and then if the actual operation is failing, switch it back to old value. Maybe that was what you meant. However, I'll let it be for now, but we could tune this later.

Reply:
Anyway, I changed this a little bit, so if we choose to use multiple threads in some point it will be cleaner. This is a little it over-engineering if we count the facts mentioned, but it's better. You are correct.

Reply:
Actually, after a good night sleep, I switched back to square one with two new powers, volatile for instance variable and running job with delay. Code is nice and clear now. :)